### PR TITLE
[V6.x] Add fallbackURLScheme support to PayPal app switch flow 

### PR DESCRIPTION
**Pull request by ndrsKloss**
### Summary of changes

We will give merchants a field to add a uri scheme for paypal to use as a backup if deeplink to merchant fails.
Changes made include

1. Adding the new parameter
2. Sending up the value
3. Checking if uri scheme value in BT validation on return.

Original PR: https://github.com/braintree/braintree_ios/pull/1673

## Videos

Setup: XOSphere with SPM pointing to my branch

Fallback with universal link activated:
https://github.com/user-attachments/assets/fe111b18-98d9-42e4-b201-6ee8078bef65

Fallback with universal link deactivated:
https://github.com/user-attachments/assets/7b95cd48-307d-4478-8478-f4b64fa40bff

### Checklist

- [x] Added a changelog entry
- [x] Tested and confirmed payment flows affected by this change are functioning as expected

- @ndrsKloss

### CODE REVIEW

File: Sources/BraintreePayPal/BTPayPalClient.swift [hunk 1]
```diff
@@ -102,13 +104,14 @@ import BraintreeDataCollector
     ///   - apiClient: The API Client
     ///   - universalLink: The URL to use for the PayPal app switch flow. Must be a valid HTTPS URL dedicated to Braintree app switch returns. This URL must be allow-listed in your Braintree Control Panel.
     /// - Warning: This initializer should be used for merchants using the PayPal App Switch flow. This feature is currently in beta and may change or be removed in future releases.
-    @objc(initWithAPIClient:universalLink:)
-    public convenience init(apiClient: BTAPIClient, universalLink: URL) {
+    @objc(initWithAPIClient:universalLink:fallbackURLScheme:)
+    public convenience init(apiClient: BTAPIClient, universalLink: URL, fallbackURLScheme: String? = nil) {
```
Comment (jaxdesmarais) [inline Sources/BraintreePayPal/BTPayPalClient.swift:109]: We are missing a docstring in the init for this param

File: Sources/BraintreePayPal/BTPayPalReturnURL.swift [hunk 2]
```diff
@@ -38,11 +38,11 @@ struct BTPayPalReturnURL {
     /// Evaluates whether the url represents a valid PayPal return URL.
     /// - Parameter url: an app switch or ASWebAuthenticationSession return URL
     /// - Returns: `true` if the url represents a valid PayPal app switch return
-    static func isValid(_ url: URL) -> Bool {
-        let isHTTPSScheme = url.scheme == "https"
-        let containsAppSwitchPath = url.path.contains("braintreeAppSwitchPayPal")
+    static func isValid(_ url: URL, fallbackURLScheme: String?) -> Bool {
+        let isValidScheme = ["https", fallbackURLScheme].contains(url.scheme)
+        let containsAppSwitchPath = url.path.contains("braintreeAppSwitchPayPal") || fallbackURLScheme == url.scheme
```
Comment (jaxdesmarais) [inline Sources/BraintreePayPal/BTPayPalReturnURL.swift]: This property is no longer doing what we say it does. We should have 2 sections here 1 for checking the `fallbackURLScheme` values and 1 for checking the HTTPS values. That way our properties can be properly named and we can do the checks as expected for each.

File: CHANGELOG.md [hunk 3]
```diff
@@ -12,6 +16,7 @@
   * Add `userAction` property to `BTPayPalVaultRequest`
   * Add `paypal:tokenize:default-browser-switch:started` event for when the appSwitch fails for universalLink and `openURLInDefaultBrowser` is triggered.
   * Add `paypal:tokenize:default-browser-switch:succeeded` and `paypal:tokenize:default-browser-switch:failed` events for when the browser switch succeeds or fails.
+  * Add `fallbackURLScheme` to `BTPayPalClient` initializer for merchants to have fallback return url if the universal link fails
```
Comment (jaxdesmarais) [inline CHANGELOG.md]: I think we can remove this one

File: Sources/BraintreePayPal/BTPayPalReturnURL.swift (hunk 2 above)
Comment (ndrsKloss) [reply Sources/BraintreePayPal/BTPayPalReturnURL.swift]: Good point! Refactored to separate the HTTPS universal link checks from the fallback URL scheme checks, with properly named variables for each section.

File: CHANGELOG.md (hunk 3 above)
Comment (ndrsKloss) [reply CHANGELOG.md]: Done

File: Sources/BraintreePayPal/BTPayPalClient.swift (hunk 1 above)
Comment (ndrsKloss) [reply Sources/BraintreePayPal/BTPayPalClient.swift:109]: Done

File: UnitTests/BraintreePayPalTests/BTPayPalCheckoutRequest_Tests.swift [hunk 4]
```diff
@@ -276,7 +276,7 @@ class BTPayPalCheckoutRequest_Tests: XCTestCase {
     func testParametersWithConfiguration_setsAppSwitchParameters_WithoutUserAuthenticationEmail() {
         let request = BTPayPalCheckoutRequest(enablePayPalAppSwitch: true, amount: "1")
         
-        let parameters = request.parameters(with: configuration, universalLink: URL(string: "some-url")!, isPayPalAppInstalled: true)
+        let parameters = request.parameters(with: configuration, universalLink: URL(string: "some-url")!, fallbackURLScheme: "paypal", isPayPalAppInstalled: true)
```
Comment (jaxdesmarais) [inline UnitTests/BraintreePayPalTests/BTPayPalCheckoutRequest_Tests.swift]: Do we need any new tests in this file similar to the Vault tests?

Comment (ndrsKloss) [reply UnitTests/BraintreePayPalTests/BTPayPalCheckoutRequest_Tests.swift]: Hey Jax, we don't need, fallbackURLScheme is only for Vault flows per the doc, so no new tests needed in Checkout.

Comment (jaxdesmarais) [reply UnitTests/BraintreePayPalTests/BTPayPalCheckoutRequest_Tests.swift]: It's added to the checkout request and as this PR is written will be available to both Checkout and Vault flows. Additionally there were tests for this added in the v7 PR. Lets check on this since as both PRs are written it can be used for both request types.

Comment (ndrsKloss) [reply UnitTests/BraintreePayPalTests/BTPayPalCheckoutRequest_Tests.swift]: Removed the additional paramter, but we can't remove it from the method signature because it overrides `BTPayPalRequest.parameters()`, which is shared by both Vault and Checkout requests.

File: Sources/BraintreePayPal/BTPayPalReturnURL.swift [hunk 5]
```diff
@@ -36,15 +36,23 @@ struct BTPayPalReturnURL {
...
-    /// - Parameter url: an app switch or ASWebAuthenticationSession return URL
+    /// - Parameters:
+    ///   - url: an app switch or ASWebAuthenticationSession return URL
+    ///   - fallbackURLScheme: Optional fallback URL scheme to check against if universal link check fails
     /// - Returns: `true` if the url represents a valid PayPal app switch return
-    static func isValid(_ url: URL) -> Bool {
-        let isHTTPSScheme = url.scheme == "https"
+    static func isValid(_ url: URL, fallbackURLScheme: String? = nil) -> Bool {
+        let containsExpectedAction = url.path.contains("cancel") || url.path.contains("success")
         let containsAppSwitchPath = url.path.contains("braintreeAppSwitchPayPal")
```
Comment (jaxdesmarais) [inline Sources/BraintreePayPal/BTPayPalReturnURL.swift:45]: Will the PayPal API be appending this portion of the path for the fallback URL use case? Right now this is appended by the SDK only for universal links unless we intended to update URL fallbacks to do the same. Right now I don't think this will pass for merchants unless we are appending this for them upstream.

File: UnitTests/BraintreePayPalTests/BTPayPalVaultRequest_Tests.swift [hunk 6]
```diff
@@ -104,12 +104,13 @@ class BTPayPalVaultRequest_Tests: XCTestCase {
...
         )
 
-        let parameters = request.parameters(with: configuration, universalLink: URL(string: "some-url")!, isPayPalAppInstalled: true)
+        let parameters = request.parameters(with: configuration, universalLink: URL(string: "some-url")!, fallbackURLScheme: "myapp", isPayPalAppInstalled: true)
 
         XCTAssertEqual(parameters["launch_paypal_app"] as? Bool, true)
         XCTAssertTrue((parameters["os_version"] as! String).matches("\\d+\\.\\d+"))
         XCTAssertTrue((parameters["os_type"] as! String).matches("iOS|iPadOS"))
         XCTAssertEqual(parameters["merchant_app_return_url"] as? String, "some-url")
+        XCTAssertEqual(parameters["merchant_app_fallback_url_scheme"] as? String, "myapp")
```
Comment (jaxdesmarais) [inline UnitTests/BraintreePayPalTests/BTPayPalVaultRequest_Tests.swift:113]: Can we add a test case that this is nil when expected?

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,9 @@
 # Braintree iOS SDK Release Notes
 
+## unreleased
+* PayPal
+  * Add optional `fallbackURLScheme` parameter to support custom URL scheme fallback in PayPal App Switch
```
Comment (jaxdesmarais) [inline CHANGELOG.md]:
```suggestion
  * Add optional `fallbackURLScheme` parameter to support custom URL scheme fallback in the PayPal App Switch flow
```

File: Sources/BraintreePayPal/BTPayPalReturnURL.swift (hunk 5 above)
Comment (ndrsKloss) [reply Sources/BraintreePayPal/BTPayPalReturnURL.swift:45]: I tested the fallback URL flow and it's working. The return URL from the PayPal backend already includes `/braintreeAppSwitchPayPal/success` in the path. When we build the fallback URL, we just swap the scheme but keep the same path. So the `isValid` check passes because `braintreeAppSwitchPayPal` is preserved in the path. Some evidences from my logs:

`> com.paypal.XOSphereApp.payments://braintree-payments/braintreeAppSwitchPayPal/success?token=..."`

File: UnitTests/BraintreePayPalTests/BTPayPalVaultRequest_Tests.swift (hunk 6 above)
Comment (ndrsKloss) [reply UnitTests/BraintreePayPalTests/BTPayPalVaultRequest_Tests.swift:113]: Done!

File: CHANGELOG.md (hunk 7 above)
Comment (ndrsKloss) [reply CHANGELOG.md]: Done!

File: Sources/BraintreePayPal/BTPayPalVaultBaseRequest.swift [hunk 8]
```diff
@@ -66,6 +72,10 @@ import BraintreeCore
 
             vaultParameters["shipping_address"] = shippingAddressParameters
         }
+        
+        if let fallbackURLScheme {
+            vaultParameters["merchant_app_fallback_url_scheme"] = fallbackURLScheme
```
Comment (jaxdesmarais) [inline Sources/BraintreePayPal/BTPayPalVaultBaseRequest.swift]: I think we discussed internally setting this on the `BTPayPalRequest`. That way when deeplinks are enabled there a second merchant upgrade isn't needed. If that's the case we may only need it on the base request class since it's set in the init, but not 100% sure.

Comment (ndrsKloss) [reply Sources/BraintreePayPal/BTPayPalVaultBaseRequest.swift]: You're correct, updting it

File: Sources/BraintreePayPal/BTPayPalClient.swift [hunk 9]
```diff
@@ -100,15 +102,17 @@ import BraintreeDataCollector
     /// Initialize a new PayPal client instance for the PayPal App Switch flow.
     /// - Parameters:
     ///   - apiClient: The API Client
-    ///   - universalLink: The URL to use for the PayPal app switch flow. Must be a valid HTTPS URL dedicated to Braintree app switch returns. This URL must be allow-listed in your Braintree Control Panel.
+    ///   - universalLink: The URL to use for the PayPal app switch flow. Must be a valid HTTPS URL dedicated to Braintree app switch returns. This URL must be allowlisted in your Braintree Control Panel.
+    ///   - fallbackURLScheme: Optional: A custom URL scheme to use as a fallback if the universal link fails. This scheme must be registered in your app's Info.plist and allowlisted in the Braintree Control Panel.
```
Comment (jaxdesmarais) [inline Sources/BraintreePayPal/BTPayPalClient.swift]: We should update this docstring to include that mercahnts _do not_ need need to include the full scheme. Most merchants will assume that this includes `://` since that's typically how URL schemes are implemented. Also is the allowlist piece correct? Is that something that needs to be done for this?

File: Sources/BraintreePayPal/BTPayPalReturnURL.swift [hunk 10]
```diff
@@ -36,15 +36,23 @@ struct BTPayPalReturnURL {
     // MARK: - Static Methods
 
     /// Evaluates whether the url represents a valid PayPal return URL.
-    /// - Parameter url: an app switch or ASWebAuthenticationSession return URL
+    /// - Parameters:
+    ///   - url: an app switch or ASWebAuthenticationSession return URL
+    ///   - fallbackURLScheme: Optional fallback URL scheme
```
Comment (agedd) [inline Sources/BraintreePayPal/BTPayPalReturnURL.swift]:
```suggestion
    ///   - url: An App Switch or `ASWebAuthenticationSession` return URL
    ///   - fallbackURLScheme: Optional. A fallback URL scheme if the universal link fails.
```

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,5 +1,9 @@
 # Braintree iOS SDK Release Notes
 
+## unreleased
+* PayPal
```
Comment (agedd) [inline CHANGELOG.md]:
```suggestion
* BraintreePayPal
```

Comment (ndrsKloss) [reply CHANGELOG.md]: Done

File: Sources/BraintreePayPal/BTPayPalReturnURL.swift (hunk 10 above)
Comment (ndrsKloss) [reply Sources/BraintreePayPal/BTPayPalReturnURL.swift]: Done

File: Sources/BraintreePayPal/BTPayPalClient.swift (hunk 9 above)
Comment (ndrsKloss) [reply Sources/BraintreePayPal/BTPayPalClient.swift]: Done, please check if it makes more sense now

> Also is the allowlist piece correct?

Yes, I will ask Semih to clarify it in the docs

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,5 +1,9 @@
 # Braintree iOS SDK Release Notes
 
+## unreleased
+* BraintreePayPal
+  * Add optional `fallbackURLScheme` parameter to support custom URL scheme fallback in PayPal App Switch flow
```
Comment (jaxdesmarais) [inline CHANGELOG.md]: Small nit: 
```suggestion
  * Add optional `fallbackURLScheme` parameter to `BTPayPalClient` to support custom URL scheme fallback in PayPal App Switch flow
```

Comment (ndrsKloss) [reply CHANGELOG.md]: Done

File: Sources/BraintreePayPal/BTPayPalClient.swift [hunk 13]
```diff
@@ -765,6 +770,6 @@ extension BTPayPalClient: BTAppContextSwitchClient {
     /// :nodoc:
     @_documentation(visibility: private)
     @objc public static func canHandleReturnURL(_ url: URL) -> Bool {
-        BTPayPalReturnURL.isValid(url)
+        BTPayPalReturnURL.isValid(url, fallbackURLScheme: payPalClient?.fallbackURLScheme)
```
Comment (jaxdesmarais) [inline Sources/BraintreePayPal/BTPayPalClient.swift:773]: small nit: I believe we should just be able to access the property directly here vs using the optional client:

```suggestion
        BTPayPalReturnURL.isValid(url, fallbackURLScheme: fallbackURLScheme)
```

File: Sources/BraintreePayPal/BTPayPalCheckoutRequest.swift [hunk 14]
```diff
@@ -168,9 +168,11 @@ import BraintreeCore
     public override func parameters(
         with configuration: BTConfiguration,
         universalLink: URL? = nil,
+        fallbackURLScheme: String? = nil,
         isPayPalAppInstalled: Bool = false
     ) -> [String: Any] {
         var baseParameters = super.parameters(with: configuration, universalLink: universalLink, isPayPalAppInstalled: isPayPalAppInstalled)
```
Comment (jaxdesmarais) [inline Sources/BraintreePayPal/BTPayPalCheckoutRequest.swift]: shouldn't we be passing `fallbackURLScheme` here?

File: Sources/BraintreePayPal/BTPayPalClient.swift [hunk 15]
```diff
@@ -100,15 +102,17 @@ import BraintreeDataCollector
     /// Initialize a new PayPal client instance for the PayPal App Switch flow.
     /// - Parameters:
     ///   - apiClient: The API Client
```
Comment (agedd) [inline Sources/BraintreePayPal/BTPayPalClient.swift:104]:
```suggestion
    ///   - apiClient: Required. The API Client
```

File: UnitTests/BraintreePayPalTests/BTPayPalVaultRequest_Tests.swift [hunk 16]
```diff
@@ -104,12 +104,26 @@ class BTPayPalVaultRequest_Tests: XCTestCase {
...
+    }
+    
+    func testParameters_withEnablePayPalAppSwitchTrue_withoutFallbackURLScheme_doesNotIncludeFallbackURLScheme() {
+        let request = BTPayPalVaultRequest(
+            userAuthenticationEmail: "sally@gmail.com",
+            enablePayPalAppSwitch: true
+        )
+
```
Comment (agedd) [inline UnitTests/BraintreePayPalTests/BTPayPalVaultRequest_Tests.swift]: nit: can remove the empty line here
```suggestion
```

File: Sources/BraintreePayPal/BTPayPalClient.swift (hunk 13 above)
Comment (ndrsKloss) [reply Sources/BraintreePayPal/BTPayPalClient.swift:773]: As `canHandleReturnURL` is a static method and `fallbackURLScheme` is an instance property, I think we need to access it through the static `payPalClient` reference. Otherwise: "Instance member 'fallbackURLScheme' cannot be used on type 'BTPayPalClient'"

File: Sources/BraintreePayPal/BTPayPalClient.swift (hunk 15 above)
Comment (ndrsKloss) [reply Sources/BraintreePayPal/BTPayPalClient.swift:104]: Is that right? We have other three clients and they don explicity use the wording `Required.`
image

File: UnitTests/BraintreePayPalTests/BTPayPalVaultRequest_Tests.swift (hunk 16 above)
Comment (ndrsKloss) [reply UnitTests/BraintreePayPalTests/BTPayPalVaultRequest_Tests.swift]: Done

File: Sources/BraintreePayPal/BTPayPalClient.swift (hunk 13 above)
Comment (jaxdesmarais) [reply Sources/BraintreePayPal/BTPayPalClient.swift:773]: Ah, gotcha. Forgot this method was static. Thanks for clarifying.

File: Sources/BraintreePayPal/BTPayPalClient.swift (hunk 15 above)
Comment (jaxdesmarais) [reply Sources/BraintreePayPal/BTPayPalClient.swift:104]: We moved to marking as required as part of v7. I would say it's helpful for merchants, but not a blocker.

File: Sources/BraintreePayPal/BTPayPalCheckoutRequest.swift (hunk 14 above)
Comment (jaxdesmarais) [reply Sources/BraintreePayPal/BTPayPalCheckoutRequest.swift]: Once `fallbackURLScheme` is added to the super here and we've added a few tests for this, we are good to go. We confirmed yesterday the GW will ignore this, but allows merchants to not need to make another update when this is released for 1-time.

Comment (ndrsKloss) [reply Sources/BraintreePayPal/BTPayPalCheckoutRequest.swift]: Changes done
